test/test_rpc_endpoints.py

```diff
@@ -0,0 +1,641 @@
+    )
+    block = resp["result"]
+    expected = {
+        "parent_hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+        "block_number": 0,
+        "status": "ACCEPTED_ON_L2",
+        "sequencer": "0x0000000000000000000000000000000000000000",
+        "old_root": "0x0",
+    }
+
+    for key, value in expected.items():
+        assert block[key] == value
+
+
+def test_get_block_by_number_raises_on_incorrect_number(deploy_info):
+    """
+    Get block by incorrect number
+    """
+    ex = rpc_call(
+        "starknet_getBlockByNumber", params={"block_number": 1234}
+    )
+
+    assert ex["error"] == {
+        "code": 26,
+        "message": "Invalid block number"
+    }
+
+
+def test_get_block_by_hash(deploy_info):
+    """
+    Get block by hash
+    """
+    block_hash = gateway_call("get_block", blockNumber=0)["block_hash"]
+
+    resp = rpc_call(
+        "starknet_getBlockByHash", params={"block_hash": block_hash}
+    )
+    block2 = resp["result"]
+    expected = {
+        "block_hash": block_hash,
+        "parent_hash": "0x0000000000000000000000000000000000000000000000000000000000000000",
+        "block_number": 0,
+        "status": "ACCEPTED_ON_L2",
+        "sequencer": "0x0000000000000000000000000000000000000000",
+        "old_root": "0x0",
+    }
+
+    for key, value in expected.items():
+        assert block2[key] == value
+
+
+def test_get_block_by_hash_raises_on_incorrect_hash(deploy_info):
+    """
+    Get block by incorrect hash
+    """
+    ex = rpc_call(
+        "starknet_getBlockByHash", params={"block_hash": "0x0"}
+    )
+
+    assert ex["error"] == {
+        "code": 24,
+        "message": "Invalid block hash"
+    }
+
+
+def test_get_storage_at(deploy_info):
+    """
+    Get storage at address
+    """
+    block = gateway_call("get_block", blockNumber=0)
+
+    contract_address: str = deploy_info["address"]
+    key: str = str(hex(get_storage_var_address("balance")))
+    block_hash: str = block["block_hash"]
+
+    resp = rpc_call(
+        "starknet_getStorageAt", params={
+            "contract_address": contract_address,
+            "key": key,
+            "block_hash": block_hash,
+        }
+    )
+    storage = resp["result"]
+
+    assert storage == "0x0"
+
+
+def test_get_storage_at_raises_on_incorrect_contract(deploy_info):
+    """
+    Get storage at incorrect contract
+    """
+    block = gateway_call("get_block", blockNumber=0)
+
+    key: str = str(hex(get_storage_var_address("balance")))
+    block_hash: str = block["block_hash"]
+
+    ex = rpc_call(
+        "starknet_getStorageAt", params={
+            "contract_address": "0x0",
+            "key": key,
+            "block_hash": block_hash,
+        }
+    )
+
+    assert ex["error"] == {
+        "code": 20,
+        "message": "Contract not found"
+    }
+
+
+# FIXME internal workings of get_storage_at would have to be changed for this to work
```

Comment:
How would you solve this by changing internal logic of feeder_gateway's get_storage_at?

Reply:
So afaik RPCs `get_storage_at` allows for this additional `block_hash` parameter so you can access storage "from the past". For this to work in devnet, internal get_storage_at would also have to accept parameter like that. 

Interestingly, I think that mainnet/testnet gateway also has `block_hash` and `block_number` parameters, but I'm unsure how they influence the returned value.

Reply:
The way get_storage_at works now is it only accepts `blockHash="latest"` which is the same as `call`. I don't think we can realistically checking storage in the past though I'm open for suggestions.